File: runtime/dart/async/zone.ts
import { IDuration } from "../core/duration";
import { IMap } from "../core/map";
import { IStackTrace } from "../core/stackTrace";
import { IAsyncError } from "./asyncError";
import { ITimer } from "./timer";
import { IZoneSpecification } from "./zoneSpecification";
declare const dart: {
    async: {
        zoneCurrent: () => IZone;
    };
};
export interface IZone {
    handleUncaughtError: (error: Object, stackTrace: IStackTrace) => void;
    getParent: () => IZone | undefined;
    getErrorZone: () => IZone;
    inSameErrorZone: (otherZone: IZone) => boolean;
    fork: (props?: {
        specification?: IZoneSpecification | undefined;
        zoneValues?: IMap<Object | undefined, Object | undefined> | undefined;
    }) => IZone;
    run: <R>(action: () => R) => R;
    runUnary: <R, T>(action: (argument: T) => R, argument: T) => R;
    runBinary: <R, T1, T2>(
        action: (argument1: T1, argument2: T2) => R,
        argument1: T1,
        argument2: T2
    ) => R;
    runGuarded: (action: () => void) => void;
    runUnaryGuarded: <T>(action: (argument: T) => void, argument: T) => void;
    runBinaryGuarded: <T1, T2>(
        action: (argument1: T1, argument2: T2) => void,
        argument1: T1,
        argument2: T2
    ) => void;
    registerCallback: <R>(callback: () => R) => () => R;
    registerUnaryCallback: <R, T>(callback: (arg: T) => R) => (arg: T) => R;
    registerBinaryCallback: <R, T1, T2>(
        callback: (arg1: T1, arg2: T2) => R
    ) => (arg1: T1, arg2: T2) => R;
    bindCallback: <R>(callback: () => R) => () => R;
    bindUnaryCallback: <R, T>(callback: (argument: T) => R) => (arg: T) => R;
    bindBinaryCallback: <R, T1, T2>(
        callback: (argument1: T1, argument2: T2) => R
    ) => (arg1: T1, arg2: T2) => R;
    bindCallbackGuarded: (callback: () => void) => () => void;
    bindUnaryCallbackGuarded: <T>(
        callback: (argument: T) => void
    ) => (__: T) => void;
    bindBinaryCallbackGuarded: <T1, T2>(
        callback: (argument1: T1, argument2: T2) => void
    ) => (__: T2) => void;
    errorCallback: (
        error: Object,
        stackTrace?: IStackTrace | undefined
    ) => IAsyncError | undefined;
    scheduleMicrotask: (callback: () => void) => void;
    createTimer: (duration: IDuration, callback: () => void) => ITimer;
    createPeriodicTimer: (
        period: IDuration,
        callback: (timer: ITimer) => void
    ) => ITimer;
    print: (line: string) => void;
}
export class Zone {
    public static current(): IZone {
        return dart.async.zoneCurrent();
    }
    private readonly _dart_handleUncaughtError: (
        error: Object,
        stackTrace: IStackTrace
    ) => void = undefined as any;
    private readonly _dart_getParent: () => IZone | undefined =
        undefined as any;
    private readonly _dart_getErrorZone: () => IZone = undefined as any;
    private readonly _dart_inSameErrorZone: (otherZone: IZone) => boolean =
        undefined as any;
    private readonly _dart_fork: (props?: {
        specification?: IZoneSpecification | undefined;
        zoneValues?: IMap<Object | undefined, Object | undefined> | undefined;
    }) => IZone = undefined as any;
    private readonly _dart_run: <R>(action: () => R) => R = undefined as any;
    private readonly _dart_runUnary: <R, T>(
        action: (argument: T) => R,
        argument: T
    ) => R = undefined as any;
    private readonly _dart_runBinary: <R, T1, T2>(
        action: (argument1: T1, argument2: T2) => R,
        argument1: T1,
        argument2: T2
    ) => R = undefined as any;
    private readonly _dart_runGuarded: (action: () => void) => void =
        undefined as any;
    private readonly _dart_runUnaryGuarded: <T>(
        action: (argument: T) => void,
        argument: T
    ) => void = undefined as any;
    private readonly _dart_runBinaryGuarded: <T1, T2>(
        action: (argument1: T1, argument2: T2) => void,
        argument1: T1,
        argument2: T2
    ) => void = undefined as any;
    private readonly _dart_registerCallback: <R>(callback: () => R) => () => R =
        undefined as any;
    private readonly _dart_registerUnaryCallback: <R, T>(
        callback: (arg: T) => R
    ) => (arg: T) => R = undefined as any;
    private readonly _dart_registerBinaryCallback: <R, T1, T2>(
        callback: (arg1: T1, arg2: T2) => R
    ) => (arg1: T1, arg2: T2) => R = undefined as any;
    private readonly _dart_bindCallback: <R>(callback: () => R) => () => R =
        undefined as any;
    private readonly _dart_bindUnaryCallback: <R, T>(
        callback: (argument: T) => R
    ) => (arg: T) => R = undefined as any;
    private readonly _dart_bindBinaryCallback: <R, T1, T2>(
        callback: (argument1: T1, argument2: T2) => R
    ) => (arg1: T1, arg2: T2) => R = undefined as any;
    private readonly _dart_bindCallbackGuarded: (
        callback: () => void
    ) => () => void = undefined as any;
    private readonly _dart_bindUnaryCallbackGuarded: <T>(
        callback: (argument: T) => void
    ) => (__: T) => void = undefined as any;
    private readonly _dart_bindBinaryCallbackGuarded: <T1, T2>(
        callback: (argument1: T1, argument2: T2) => void
    ) => (__: T2) => void = undefined as any;
    private readonly _dart_errorCallback: (
        error: Object,
        stackTrace?: IStackTrace | undefined
    ) => IAsyncError | undefined = undefined as any;
    private readonly _dart_scheduleMicrotask: (callback: () => void) => void =
        undefined as any;
    private readonly _dart_createTimer: (
        duration: IDuration,
        callback: () => void
    ) => ITimer = undefined as any;
    private readonly _dart_createPeriodicTimer: (
        period: IDuration,
        callback: (timer: ITimer) => void
    ) => ITimer = undefined as any;
    private readonly _dart_print: (line: string) => void = undefined as any;
    public handleUncaughtError(error: Object, stackTrace: IStackTrace): void {
        return this._dart_handleUncaughtError(error, stackTrace);
    }
    public getParent(): IZone | undefined {
        return this._dart_getParent();
    }
    public getErrorZone(): IZone {
        return this._dart_getErrorZone();
    }
    public inSameErrorZone(otherZone: IZone): boolean {
        return this._dart_inSameErrorZone(otherZone);
    }
    public fork(props?: {
        specification?: IZoneSpecification | undefined;
        zoneValues?: IMap<Object | undefined, Object | undefined> | undefined;
    }): IZone {
        return this._dart_fork(props);
    }
    public run<R>(action: () => R): R {
        return this._dart_run(action);
    }
    public runUnary<R, T>(action: (argument: T) => R, argument: T): R {
        return this._dart_runUnary(action, argument);
    }
    public runBinary<R, T1, T2>(
        action: (argument1: T1, argument2: T2) => R,
        argument1: T1,
        argument2: T2
    ): R {
        return this._dart_runBinary(action, argument1, argument2);
    }
    public runGuarded(action: () => void): void {
        return this._dart_runGuarded(action);
    }
    public runUnaryGuarded<T>(
        action: (argument: T) => void,
        argument: T
    ): void {
        return this._dart_runUnaryGuarded(action, argument);
    }
    public runBinaryGuarded<T1, T2>(
        action: (argument1: T1, argument2: T2) => void,
        argument1: T1,
        argument2: T2
    ): void {
        return this._dart_runBinaryGuarded(action, argument1, argument2);
    }
    public registerCallback<R>(callback: () => R): () => R {
        return this._dart_registerCallback(callback);
    }
    public registerUnaryCallback<R, T>(callback: (arg: T) => R): (arg: T) => R {
        return this._dart_registerUnaryCallback(callback);
    }
    public registerBinaryCallback<R, T1, T2>(
        callback: (arg1: T1, arg2: T2) => R
    ): (arg1: T1, arg2: T2) => R {
        return this._dart_registerBinaryCallback(callback);
    }
    public bindCallback<R>(callback: () => R): () => R {
        return this._dart_bindCallback(callback);
    }
    public bindUnaryCallback<R, T>(
        callback: (argument: T) => R
    ): (arg: T) => R {
        return this._dart_bindUnaryCallback(callback);
    }
    public bindBinaryCallback<R, T1, T2>(
        callback: (argument1: T1, argument2: T2) => R
    ): (arg1: T1, arg2: T2) => R {
        return this._dart_bindBinaryCallback(callback);
    }
    public bindCallbackGuarded(callback: () => void): () => void {
        return this._dart_bindCallbackGuarded(callback);
    }
    public bindUnaryCallbackGuarded<T>(
        callback: (argument: T) => void
    ): (__: T) => void {
        return this._dart_bindUnaryCallbackGuarded(callback);
    }
    public bindBinaryCallbackGuarded<T1, T2>(
        callback: (argument1: T1, argument2: T2) => void
    ): (__: T2) => void {
        return this._dart_bindBinaryCallbackGuarded(callback);
    }
    public errorCallback(
        error: Object,
        stackTrace?: IStackTrace | undefined
    ): IAsyncError | undefined {
        return this._dart_errorCallback(error, stackTrace);
    }
    public scheduleMicrotask(callback: () => void): void {
        return this._dart_scheduleMicrotask(callback);
    }
    public createTimer(duration: IDuration, callback: () => void): ITimer {
        return this._dart_createTimer(duration, callback);
    }
    public createPeriodicTimer(
        period: IDuration,
        callback: (timer: ITimer) => void
    ): ITimer {
        return this._dart_createPeriodicTimer(period, callback);
    }
    public print(line: string): void {
        return this._dart_print(line);
    }
}

File: lib/cfr/builtins/libs/dart/async/zone.dart
import 'dart:async';
import 'dart:core';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedZone extends VMManagedBox<Zone> {
  VMManagedZone(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['handleUncaughtError'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.handleUncaughtError(
          maybeUnBoxAndBuildArgument<Object, dynamic>(luaCallerArguments[1],
              parentState: hydroState),
          maybeUnBoxAndBuildArgument<StackTrace, dynamic>(luaCallerArguments[2],
              parentState: hydroState));
      return [];
    });
    table['getParent'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      final returnValue = vmObject.parent;
      if (returnValue != null) {
        return [
          maybeBoxObject<Zone?>(
              object: returnValue, hydroState: hydroState, table: HydroTable()),
        ];
      }
      return [];
    });
    table['getErrorZone'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Zone>(
            object: vmObject.errorZone,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['inSameErrorZone'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.inSameErrorZone(maybeUnBoxAndBuildArgument<Zone, dynamic>(
            luaCallerArguments[1],
            parentState: hydroState)),
      ];
    });
    table['fork'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Zone>(
            object: vmObject.fork(
                specification:
                    maybeUnBoxAndBuildArgument<ZoneSpecification?, dynamic>(
                        luaCallerArguments.length >= 2
                            ? luaCallerArguments[1]['specification']
                            : null,
                        parentState: hydroState),
                zoneValues:
                    maybeUnBoxAndBuildArgument<Map<Object?, Object?>?, Object?>(
                        luaCallerArguments.length >= 2
                            ? luaCallerArguments[1]['zoneValues']
                            : null,
                        parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['run'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedaction = luaCallerArguments[1];
      return [
        vmObject.run(() => unpackedaction.dispatch(
              [
                luaCallerArguments[0],
              ],
              parentState: hydroState,
            )[0]),
      ];
    });
    table['runUnary'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedaction = luaCallerArguments[1];
      return [
        vmObject.runUnary(
            (argument) => unpackedaction.dispatch(
                  [luaCallerArguments[0], argument],
                  parentState: hydroState,
                )[0],
            luaCallerArguments[2]),
      ];
    });
    table['runBinary'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedaction = luaCallerArguments[1];

      return [
        vmObject.runBinary(
            (argument1, argument2) => unpackedaction.dispatch(
                  [luaCallerArguments[0], argument1, argument2],
                  parentState: hydroState,
                )[0],
            luaCallerArguments[2],
            luaCallerArguments[3]),
      ];
    });
    table['runGuarded'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedaction = luaCallerArguments[1];
      vmObject.runGuarded(() => unpackedaction.dispatch(
            [
              luaCallerArguments[0],
            ],
            parentState: hydroState,
          ));
      return [];
    });
    table['runUnaryGuarded'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedaction = luaCallerArguments[1];
      vmObject.runUnaryGuarded(
          (argument) => unpackedaction.dispatch(
                [luaCallerArguments[0], argument],
                parentState: hydroState,
              ),
          luaCallerArguments[2]);
      return [];
    });
    table['runBinaryGuarded'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedaction = luaCallerArguments[1];

      vmObject.runBinaryGuarded(
          (argument1, argument2) => unpackedaction.dispatch(
                [luaCallerArguments[0], argument1, argument2],
                parentState: hydroState,
              ),
          luaCallerArguments[2],
          luaCallerArguments[3]);
      return [];
    });
    table['registerCallback'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[1];
      return [
        vmObject.registerCallback(() => unpackedcallback.dispatch(
              [
                luaCallerArguments[0],
              ],
              parentState: hydroState,
            )[0]),
      ];
    });
    table['registerUnaryCallback'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[1];
      return [
        vmObject.registerUnaryCallback((arg) => unpackedcallback.dispatch(
              [luaCallerArguments[0], arg],
              parentState: hydroState,
            )[0]),
      ];
    });
    table['registerBinaryCallback'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[1];
      return [
        vmObject
            .registerBinaryCallback((arg1, arg2) => unpackedcallback.dispatch(
                  [luaCallerArguments[0], arg1, arg2],
                  parentState: hydroState,
                )[0]),
      ];
    });
    table['bindCallback'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[1];
      return [
        vmObject.bindCallback(() => unpackedcallback.dispatch(
              [
                luaCallerArguments[0],
              ],
              parentState: hydroState,
            )[0]),
      ];
    });
    table['bindUnaryCallback'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[1];
      return [
        vmObject.bindUnaryCallback((argument) => unpackedcallback.dispatch(
              [luaCallerArguments[0], argument],
              parentState: hydroState,
            )[0]),
      ];
    });
    table['bindBinaryCallback'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[1];
      return [
        vmObject.bindBinaryCallback(
            (argument1, argument2) => unpackedcallback.dispatch(
                  [luaCallerArguments[0], argument1, argument2],
                  parentState: hydroState,
                )[0]),
      ];
    });
    table['bindCallbackGuarded'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[1];
      return [
        vmObject.bindCallbackGuarded(() => unpackedcallback.dispatch(
              [
                luaCallerArguments[0],
              ],
              parentState: hydroState,
            )),
      ];
    });
    table['bindUnaryCallbackGuarded'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[1];
      return [
        vmObject
            .bindUnaryCallbackGuarded((argument) => unpackedcallback.dispatch(
                  [luaCallerArguments[0], argument],
                  parentState: hydroState,
                )),
      ];
    });
    table['bindBinaryCallbackGuarded'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[1];
      return [
        vmObject.bindBinaryCallbackGuarded(
            (argument1, argument2) => unpackedcallback.dispatch(
                  [luaCallerArguments[0], argument1, argument2],
                  parentState: hydroState,
                )),
      ];
    });
    table['errorCallback'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      final returnValue = vmObject.errorCallback(
          maybeUnBoxAndBuildArgument<Object, dynamic>(luaCallerArguments[1],
              parentState: hydroState),
          maybeUnBoxAndBuildArgument<StackTrace?, dynamic>(
              luaCallerArguments[2],
              parentState: hydroState));
      if (returnValue != null) {
        return [
          maybeBoxObject<AsyncError?>(
              object: returnValue, hydroState: hydroState, table: HydroTable()),
        ];
      }
      return [];
    });
    table['scheduleMicrotask'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[1];
      vmObject.scheduleMicrotask(() => unpackedcallback.dispatch(
            [
              luaCallerArguments[0],
            ],
            parentState: hydroState,
          ));
      return [];
    });
    table['createTimer'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[2];
      return [
        maybeBoxObject<Timer>(
            object: vmObject.createTimer(
                maybeUnBoxAndBuildArgument<Duration, dynamic>(
                    luaCallerArguments[1],
                    parentState: hydroState),
                () => unpackedcallback.dispatch(
                      [
                        luaCallerArguments[0],
                      ],
                      parentState: hydroState,
                    )),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['createPeriodicTimer'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      Closure unpackedcallback = luaCallerArguments[2];
      return [
        maybeBoxObject<Timer>(
            object: vmObject.createPeriodicTimer(
                maybeUnBoxAndBuildArgument<Duration, dynamic>(
                    luaCallerArguments[1],
                    parentState: hydroState),
                (timer) => unpackedcallback.dispatch(
                      [luaCallerArguments[0], timer],
                      parentState: hydroState,
                    )),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['print'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      vmObject.print(luaCallerArguments[1]);
      return [];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final Zone vmObject;
}

void loadZone({required HydroState hydroState, required HydroTable table}) {
  table['zoneCurrent'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Zone>(
          object: Zone.current, hydroState: hydroState, table: HydroTable()),
    ];
  });
  registerBoxer<Zone>(boxer: (
      {required Zone vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedZone(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}
